### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by AndersMad** [Taritsyn/DartSassHost#12]
I did a quick performance test on LibSassHost (c++) vs DartSassHost (js):

CSS output size: 43kb - here the time is a little over double and as one would expect.
LibSass: 0.3s
DartSass: 0.7s

CSS output size: 131kb - here it's over 10 time as slow although the output size is only 3 times as big.
LibSass: 1.6s
DartSass: 17.2s

I suspect that the big one is so much slower because of more calls to file manager. I tried with `DisableGlobalMembers = true` and `DisableDynamicBinding = true` but no difference.

This is not an issue post as such as the problem can be Dart or V8JsEngine via ClearScript.V8 - but maybe you have a suggestion for a tweak? Else just close and thank you for your wonderful contributions! :)

**Comment by Taritsyn** [comment]
Hello, Anders!

DartSassHost will never faster than an LibSassHost. The point here is not only in the file manager, but in huge amount of code that JS engine must execute. This is largest codebase I've ever had to run on a JavaScript Engine Switcher. Moreover, I tried to greatly reduce this code base.

I recommend that you experiment with the [JavaScriptEngineSwitcher.ChakraCore](https://github.com/Taritsyn/JavaScriptEngineSwitcher/wiki/ChakraCore) module.

**Comment by AndersMad** [comment]
But note that the first one is only 2 x as slow - spite executing the entire Dart JS - that's why I wrote that is to be expected, but the next one (but a bigger pile of SCSS very fragmented into a lot of files) takes 10 times as much.

So I was thinking its all the file manager "reflection" call backs that makes it slow. Might be solved by ClearScript some day - I bet thats the culprit.

Ok, I thought that ClearScript would be the fastest / best as its the "new MS way" and that Charkra is deprecated.

But just close this one - was more a general FYI :)

**Comment by AndersMad** [comment]
Just a quick FYI update:
- I tried with ChakraCoreJs and that takes about a minute (about 3 times as long as V8).  
- I also tried YantraJs - but that throws an error ("Unexpected token Identifier" @ sass-bundled.min.js:87:151).

**Comment by Taritsyn** [comment]
> * I also tried YantraJs - but that throws an error ("Unexpected token Identifier" @ sass-bundled.min.js:87:151).

I did not add a Yantra to the list of supported JS engines. At the moment, Yantra is not yet suitable for running JS libraries.

**Comment by AndersMad** [comment]
Yes I saw the list - but just wanted to give it a swing since you recently updated it, thanks! :)

**Comment by jzabroski** [comment]
Also worth noting I tried Jint and it throws a ton of exceptions in debug mode. No way that would be faster.

**Comment by Taritsyn** [comment]
Hello, John!

> No way that would be faster.

At one time, there was a similar problem with the performance of the [Ruby Sass](https://sass-lang.com/ruby-sass/) library, which was solved by creating the [LibSass](https://github.com/sass/libsass) library. LibSass is now deprecated and the [Dart Sass](https://github.com/sass/dart-sass) is recommended instead.

In my library I use a transpiled version of the Dart Sass from the [sass](https://www.npmjs.com/package/sass) package. I replace the Node.js specific API calls in the `sass.dart.js` file with polyfills and calls to an instance of the `IFileManager` implementation. I also remove some redundant code from this file. After cleaning, the size of this file is comparable to the size of the TypeScript compiler file. In this case, the reason for the low performance is similar to the reason that the TypeScript compiler has begun to be rewritten in Go.

Previously, I recommended using the JavaScriptEngineSwitcher.ChakraCore module to get high performance, but now I recommend a JavaScriptEngineSwitcher.V8 module.

> Also worth noting I tried Jint and it throws a ton of exceptions in debug mode. 

What kind of errors occur? It's just that I usually use the JavaScriptEngineSwitcher.Msie module in the `ChakraEdgeJsRt` mode for debugging.

In general, DartSass Host runs under the Jint without errors. TypeScript compiler does have problems with the latest versions of Jint.
